### ISSUE TITLE
Generic cache class

### ISSUE THREAD

**Issue by ddemidov** [rhempel/ev3dev-lang-python#29]
@rhempel started the discussion in #28, I think it deserves an issue of its own.

A couple of classes now have their own implementation of cache (basically just a dictionary). We could introduce generic class for cache. What bugs me is `Device` and `Button` use their caches a bit differently.

[`Device`](https://github.com/rhempel/ev3dev-lang-python/blob/68eab43b10740adb8730c2939d46e12f83e9cfc1/ev3dev.py#L109-L123):
```py
    def _attribute_file( self, attribute, mode, reopen=False ):
        """Manages the file handle cache and opening the files in the correct mode"""

        attribute_name = os.path.abspath( self._path + '/' + attribute )

        if attribute_name not in self.filehandle_cache:
            f = open( attribute_name, mode )
            self.filehandle_cache[attribute_name] = f
        elif reopen == True:
            self.filehandle_cache[attribute_name].close()
            f = open( attribute_name, mode )
            self.filehandle_cache[attribute_name] = f
        else:
            f = self.filehandle_cache[attribute_name]
        return f
```
[`Button`](https://github.com/rhempel/ev3dev-lang-python/blob/68eab43b10740adb8730c2939d46e12f83e9cfc1/ev3dev.py#L1796-L1801):
```py
    def _button_file(self, name):
        if name not in self.filehandle_cache:
            f = open( name, 'r' )
            self.filehandle_cache[name] = f
        else:
            f = self.filehandle_cache[name]
        return f
```

What should we move into the cache class? Just some key-value store (similar to dict), but with some added functionality like size limitation (like LRU cache)? Now, since each object instance has its own cache, there is probably no need to deal with size limitation.

**Comment by rhempel** [comment]
This is a file handle cache - the purpose of the extra bits in the device version is that it allows you to open a file with only the mode you ask for - maybe we can modify things so that if an attribute is r/w you get the right mode the first time :-) The reopen is only needed if you write to an attribute, and then read it later.

**Comment by ddemidov** [comment]
We can also make a key to be `file_name`/`mode`. This will have two versions of the same file in cache, but would not require reopen.

**Comment by rhempel** [comment]
I'm not sure having one descriptor open for read and another for write is a good thing - I guess as long as the seek points are truly independent it will work - not sure though.

**Comment by ddemidov** [comment]
It certainly works in C++ bindings. I agree it sounds a bit wasteful. And the trick with reopening on exception should work for all cases.

**Comment by ddemidov** [comment]
Although it's not clear how to incorporate the trick into cache, since the try block should be around both cache access and file operation. May be we should think not about file handle cache, but file reader /writer?

**Comment by rhempel** [comment]
+1 File reader/writer (that takes advantage of a cache for file handles) :-)

**Comment by ddemidov** [comment]
Something like this: https://github.com/ddemidov/ev3dev-lang-python-1/commit/92828ecf6061b8e5f0a65664eb86082874b5ee89?

**Comment by ddemidov** [comment]
Resolved by #31
